### PR TITLE
fix: Don't error if no new wheels were built

### PR DESCRIPTION
**Pull request by pauldmccarthy**
Howdy, this PR contains a simple fix for an issue which I came across when using the `--allow-empty` flag. There is some logic in `cibuildwheel.util.print_new_wheels` which is raising an error if no new wheels were built. 

```
mkdir testproj && cd testproj && touch setup.py
CIBW_BUILD="don't match anything" cibuildwheel --platform=linux --allow-empty
```

<details>

```
     _ _       _ _   _       _           _
 ___|_| |_ _ _|_| |_| |_ _ _| |_ ___ ___| |
|  _| | . | | | | | . | | | |   | -_| -_| |
|___|_|___|___|_|_|___|_____|_|_|___|___|_|

cibuildwheel version 2.4.0

Build options:
  platform: 'linux'
  architectures: {<Architecture.i686: 'i686'>, <Architecture.x86_64: 'x86_64'>}
  build_selector: BuildSelector(skip_config='', build_config="don't match anything", requires_python=None, prerelease_pythons=False)
  output_dir: PosixPath('wheelhouse')
  package_dir: PosixPath('.')
  test_selector: TestSelector(skip_config='', build_config='*', requires_python=None, prerelease_pythons=False)
  before_all: ''
  before_build: ''
  before_test: ''
  build_frontend: 'pip'
  build_verbosity: 0
  dependency_constraints: DependencyConstraints(PosixPath('/home/paulmc/Projects/cibuildwheel/cibuildwheel/resources/constraints.txt'))
  environment: ParsedEnvironment([])
  manylinux_images: {'x86_64': 'quay.io/pypa/manylinux2014_x86_64:2022-04-03-da6ecb3', 'i686': 'quay.io/pypa/manylinux2014_i686:2022-04-03-da6ecb3', 'pypy_x86_64': 'quay.io/pypa/manylinux2014_x86_64:2022-04-03-da6ecb3', 'aarch64': 'quay.io/pypa/manylinux2014_aarch64:2022-04-03-da6ecb3', 'ppc64le': 'quay.io/pypa/manylinux2014_ppc64le:2022-04-03-da6ecb3', 's390x': 'quay.io/pypa/manylinux2014_s390x:2022-04-03-da6ecb3', 'pypy_aarch64': 'quay.io/pypa/manylinux2014_aarch64:2022-04-03-da6ecb3', 'pypy_i686': 'quay.io/pypa/manylinux2014_i686:2022-04-03-da6ecb3'}
  musllinux_images: {'x86_64': 'quay.io/pypa/musllinux_1_1_x86_64:2022-04-03-da6ecb3', 'i686': 'quay.io/pypa/musllinux_1_1_i686:2022-04-03-da6ecb3', 'aarch64': 'quay.io/pypa/musllinux_1_1_aarch64:2022-04-03-da6ecb3', 'ppc64le': 'quay.io/pypa/musllinux_1_1_ppc64le:2022-04-03-da6ecb3', 's390x': 'quay.io/pypa/musllinux_1_1_s390x:2022-04-03-da6ecb3'}
  repair_command: 'auditwheel repair -w {dest_dir} {wheel}'
  test_command: ''
  test_extras: ''
  test_requires: []
Cache folder: /home/paulmc/.cache/cibuildwheel

Here we go!

cibuildwheel: No build identifiers selected: BuildSelector(skip_config='', build_config="don't match anything", requires_python=None, prerelease_pythons=False)
Traceback (most recent call last):
  File "/usr/lib/python3.8/runpy.py", line 194, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/usr/lib/python3.8/runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "/home/paulmc/Projects/cibuildwheel/cibuildwheel/__main__.py", line 295, in <module>
    main()
  File "/home/paulmc/Projects/cibuildwheel/cibuildwheel/__main__.py", line 213, in main
    assert_never(platform)
  File "/usr/lib/python3.8/contextlib.py", line 120, in __exit__
    next(self.gen)
  File "/home/paulmc/Projects/cibuildwheel/cibuildwheel/util.py", line 433, in print_new_wheels
    max_name_len = max(len(f.name) for f in new_contents)
ValueError: max() arg is an empty sequence

✗ echo $?
1
```

</details>

With this PR:

```
CIBW_BUILD="don't match anything" cibuildwheel --platform=linux --allow-empty
```

<details>

```
     _ _       _ _   _       _           _
 ___|_| |_ _ _|_| |_| |_ _ _| |_ ___ ___| |
|  _| | . | | | | | . | | | |   | -_| -_| |
|___|_|___|___|_|_|___|_____|_|_|___|___|_|

cibuildwheel version 2.4.0

Build options:
  platform: 'linux'
  architectures: {<Architecture.x86_64: 'x86_64'>, <Architecture.i686: 'i686'>}
  build_selector: BuildSelector(skip_config='', build_config="don't match anything", requires_python=None, prerelease_pythons=False)
  output_dir: PosixPath('wheelhouse')
  package_dir: PosixPath('.')
  test_selector: TestSelector(skip_config='', build_config='*', requires_python=None, prerelease_pythons=False)
  before_all: ''
  before_build: ''
  before_test: ''
  build_frontend: 'pip'
  build_verbosity: 0
  dependency_constraints: DependencyConstraints(PosixPath('/home/paulmc/Projects/cibuildwheel/cibuildwheel/resources/constraints.txt'))
  environment: ParsedEnvironment([])
  manylinux_images: {'x86_64': 'quay.io/pypa/manylinux2014_x86_64:2022-04-03-da6ecb3', 'i686': 'quay.io/pypa/manylinux2014_i686:2022-04-03-da6ecb3', 'pypy_x86_64': 'quay.io/pypa/manylinux2014_x86_64:2022-04-03-da6ecb3', 'aarch64': 'quay.io/pypa/manylinux2014_aarch64:2022-04-03-da6ecb3', 'ppc64le': 'quay.io/pypa/manylinux2014_ppc64le:2022-04-03-da6ecb3', 's390x': 'quay.io/pypa/manylinux2014_s390x:2022-04-03-da6ecb3', 'pypy_aarch64': 'quay.io/pypa/manylinux2014_aarch64:2022-04-03-da6ecb3', 'pypy_i686': 'quay.io/pypa/manylinux2014_i686:2022-04-03-da6ecb3'}
  musllinux_images: {'x86_64': 'quay.io/pypa/musllinux_1_1_x86_64:2022-04-03-da6ecb3', 'i686': 'quay.io/pypa/musllinux_1_1_i686:2022-04-03-da6ecb3', 'aarch64': 'quay.io/pypa/musllinux_1_1_aarch64:2022-04-03-da6ecb3', 'ppc64le': 'quay.io/pypa/musllinux_1_1_ppc64le:2022-04-03-da6ecb3', 's390x': 'quay.io/pypa/musllinux_1_1_s390x:2022-04-03-da6ecb3'}
  repair_command: 'auditwheel repair -w {dest_dir} {wheel}'
  test_command: ''
  test_extras: ''
  test_requires: []
Cache folder: /home/paulmc/.cache/cibuildwheel

Here we go!

cibuildwheel: No build identifiers selected: BuildSelector(skip_config='', build_config="don't match anything", requires_python=None, prerelease_pythons=False)
No new wheels

✗ echo $?
0
```

</details>

Let me know if you'd like anything adjusted (e.g. the message that is printed). I'm also happy to add an integration test if you could point me to the right place to add it.

Thanks!

### CODE REVIEW

File: cibuildwheel/util.py [hunk 1]
```diff
@@ -430,6 +430,11 @@ class FileReport(NamedTuple):
         FileReport(wheel.name, f"{(wheel.stat().st_size + 1023) // 1024:,d}")
         for wheel in final_contents - existing_contents
     ]
+
+    if len(new_contents) == 0:
+        print("No new wheels")
```
Comment (joerick) [inline cibuildwheel/util.py:435]:
```suggestion
```

Let's remove this message as there's already a message printed when no build identifiers are selected.

**Review by joerick — COMMENTED** [review]
Thanks. A test would also be good, you could add an integration test in the test_0_basic module.

**Comment by pauldmccarthy** [comment]
Hi @joerick, I'm not entirely sure of the cause of this test failure - any thoughts?
 
https://app.travis-ci.com/github/pypa/cibuildwheel/jobs/567099381#L339-L359

**Comment by joerick** [comment]
Looks like a flaky test/random failure. I've restarted it.

File: cibuildwheel/macos.py [hunk 2]
```diff
@@ -276,17 +276,20 @@ def build(options: Options, tmp_path: Path) -> None:
...
-
-        if before_all_options.before_all:
-            log.step("Running before_all...")
-            env = before_all_options.environment.as_dictionary(prev_environment=os.environ)
-            env.setdefault("MACOSX_DEPLOYMENT_TARGET", "10.9")
-            before_all_prepared = prepare_command(
-                before_all_options.before_all, project=".", package=before_all_options.package_dir
-            )
-            shell(before_all_prepared, env=env)
+        if len(python_configurations) > 0:
```
Comment (joerick) [inline cibuildwheel/macos.py:279]: Can we convert these blocks into early exits instead? Would avoid the extra indentation.

File: cibuildwheel/windows.py [hunk 3]
```diff
@@ -235,16 +235,17 @@ def build(options: Options, tmp_path: Path) -> None:
...
-        before_all_options = options.build_options(before_all_options_identifier)
-
-        if before_all_options.before_all:
-            log.step("Running before_all...")
-            env = before_all_options.environment.as_dictionary(prev_environment=os.environ)
-            before_all_prepared = prepare_command(
-                before_all_options.before_all, project=".", package=options.globals.package_dir
-            )
-            shell(before_all_prepared, env=env)
+        if len(python_configurations) > 0:
```
Comment (joerick) [inline cibuildwheel/windows.py:238]: Same here, early exit would be neater I think.

File: test/utils.py [hunk 4]
```diff
@@ -77,7 +83,8 @@ def cibuildwheel_run(project_path, package_dir=".", env=None, add_env=None, outp
                 "--output-dir",
                 str(output_dir or tmp_output_dir),
                 str(package_dir),
-            ],
+            ]
+            + add_args,
```
Comment (joerick) [inline test/utils.py:87]: A very minor thing, but it would fit the style of the codebase a bit more to do `*add_args` inside the list.

Comment (joerick) [inline test/utils.py:87]: A very minor thing, but it would fit the style of the codebase a bit more to do `*add_args` inside the list.

File: test/test_0_basic.py [hunk 5]
```diff
@@ -59,3 +59,19 @@ def test_build_identifiers(tmp_path):
     assert len(expected_wheels) == len(
         build_identifiers
     ), f"{expected_wheels} vs {build_identifiers}"
+
+
+def test_allow_empty(tmp_path, build_frontend_env):
```
Comment (joerick) [inline test/test_0_basic.py:64]: We can remove the build_frontend_env fixture from this test, it's not relevant.

**Review by joerick — COMMENTED** [review]
Brilliant, thanks! A few things to tweak and we can get this in 🙂

**Comment by pauldmccarthy** [comment]
@joerick all requested changes made - over to you!

File: cibuildwheel/macos.py [hunk 6]
```diff
@@ -284,7 +287,9 @@ def build(options: Options, tmp_path: Path) -> None:
             env = before_all_options.environment.as_dictionary(prev_environment=os.environ)
             env.setdefault("MACOSX_DEPLOYMENT_TARGET", "10.9")
             before_all_prepared = prepare_command(
-                before_all_options.before_all, project=".", package=before_all_options.package_dir
+                before_all_options.before_all,
+                project=".",
+                package=before_all_options.package_dir,
```
Comment (joerick) [inline cibuildwheel/macos.py:292]: Would you mind to revert this change? Thanks.

**Review by joerick — APPROVED** [review]
Looks good. Thank you!

File: cibuildwheel/macos.py (hunk 6 above)
Comment (henryiii) [reply cibuildwheel/macos.py:292]: Be sure to avoid adding a trailing comma. That tells black you want it one per line.

Comment (pauldmccarthy) [reply cibuildwheel/macos.py]: My mistake, fixed!
